server/src/main/java/io/unitycatalog/server/persist/utils/UriUtils.java

```diff
@@ -0,0 +1,245 @@
+package io.unitycatalog.server.persist.utils;
+
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicSessionCredentials;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import io.unitycatalog.server.exception.BaseException;
+import io.unitycatalog.server.exception.ErrorCode;
+import io.unitycatalog.server.model.AwsCredentials;
+import io.unitycatalog.server.model.AzureUserDelegationSAS;
+import io.unitycatalog.server.model.GcpOauthToken;
+import io.unitycatalog.server.model.GenerateTemporaryModelVersionCredentialsResponse;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileVisitOption;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UriUtils {
+  private static final Logger LOGGER = LoggerFactory.getLogger(UriUtils.class);
+  private static final ServerPropertiesUtils properties = ServerPropertiesUtils.getInstance();
+
+  private enum Operation {
+    CREATE,
+    DELETE
+  }
+
+  private UriUtils() {}
+
+  // Model specific storage root handlers and convenience methods
+  private static String getModelStorageRoot() {
+    String modelStorageRoot = properties.getProperty("storage-root.models");
+    if (modelStorageRoot == null) {
+      // If the model storage root is empty, use the CWD
+      modelStorageRoot = System.getProperty("user.dir");
+    }
+    // If the model storage root is not a valid URI, make it one
+    if (!isValidURI(modelStorageRoot)) {
```

Comment:
So the only reason we are assuming it's not valid is because it's relative. As pointed below, there could be genuine reason of the storage root not being valid (normalized uri prefix != uri prefix) .